### PR TITLE
Some ways to avoid seg-fault in Nim std lib

### LINKED ISSUE #10765: Bug in setTerminate()
### Example
https://github.com/bio-nim/pbbam/blob/f5a659e9fbed4b675f4458734fdfc9fe8162b2ac/nim/foo.nim

### Current Output
https://forum.nim-lang.org/t/4686
```
Traceback (most recent call last)
foo.nim(55)              foo
foo.nim(49)              main
system.nim(4341)         :anonymous
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
```
```
queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS
```
### Expected Output
We should see the exception message, similar to the default g++ terminate-handler.

### Possible Solution
I have added prints into `lib/system.nim::setTerminate()` in order to find where the crash actually occurs. The problem is that `let ex = getCurrentException()` is `nil`, so writing `ex.name` or `ex.msg` fails.

It's `nil` because nothing called `setCurrentException`.

I might be able to call `setTerminate()` (which is just`std::set_terminate()`) to install my own exception-trapping handler, as shown [here](https://stackoverflow.com/questions/3774316/c-unhandled-exceptions):
```cpp
   void terminate_handler()
   {
       try { throw; }
       catch(const std::exception& e) { log(e.what()); }
       catch(...) {}
   }
```
But I think this is a real bug anyway. Minimally, the anon proc given to `setTerminate()` should at least test `nil == ex`.

Ideally, we want _both_ the C++ exception _and_ the Nim stack-trace. Maybe a good approach is for Nim's default terminate-handler to show the Nim stack-trace first, and then to call `__gnu_cxx::__verbose_terminate_handler` if the Nim exception is `nil` and the backend is  `cpp` and GNU.

Even fancier, the C++ stack-trace is [harder to obtain](https://www.gnu.org/software/libc/manual/html_node/Backtraces.html), but that could be helpful too. (Would you like a pull-request for that, at least with GNU g++?)

### Additional Information
* Your Nim version:
  * `0.19.4`
* Was it working in the previous Nim releases?
  * Probably not.
* A link to a related issue or discussion.
  * https://gcc.gnu.org/onlinedocs/libstdc++/manual/termination.html
  * https://stackoverflow.com/questions/3774316/c-unhandled-exceptions
  * https://github.com/llvm-mirror/libcxxabi/blob/master/src/cxa_exception.cpp (to see how LLVN works)
  * https://github.com/llvm-mirror/libcxxabi/blob/master/src/cxa_handlers.cpp
  * https://github.com/gcc-mirror/gcc/blob/41d6b10e96a1de98e90a7c0378437c3255814b16/libstdc%2B%2B-v3/libsupc%2B%2B/vterminate.cc#L37

### PR DESCRIPTION
**Pull request by pb-cdunn**
fixes #10765

### CODE REVIEW

File: lib/system/excpt.nim [hunk 1]
```diff
@@ -474,12 +474,34 @@ when defined(endb):
...
+  proc getTerminate: proc() {.noconv.}
+    {.importc: "std::get_terminate", header: "<exception>".} # C++11
+  var oldTerminate = getTerminate()
   proc setTerminate(handler: proc() {.noconv.})
     {.importc: "std::set_terminate", header: "<exception>".}
   setTerminate proc() {.noconv.} =
     # Remove ourself as a handler, reinstalling the default handler.
     setTerminate(nil)
 
+    if nil == currException:
```
Comment (Araq) [inline lib/system/excpt.nim]: Nim is not C++, we happily use ``currException == nil``.

File: lib/system/excpt.nim [hunk 2]
```diff
@@ -474,12 +474,34 @@ when defined(endb):
...
+            } catch (const std::exception& exc) {
+              fprintf(stderr, "  Uncaught exception. what(): '%s'\n", exc.what());
+            } catch (...) {
+              fprintf(stderr, "  Uncaught unknown exception.\n");
+            }
+        """.}
+        return
+      else:
+        # verboseTeminate()
+        if nil != oldTerminate:
```
Comment (Araq) [inline lib/system/excpt.nim]: Same.

File: lib/system/excpt.nim [hunk 3]
```diff
@@ -474,12 +474,34 @@ when defined(endb):
...
   proc setTerminate(handler: proc() {.noconv.})
     {.importc: "std::set_terminate", header: "<exception>".}
   setTerminate proc() {.noconv.} =
     # Remove ourself as a handler, reinstalling the default handler.
     setTerminate(nil)
 
+    if nil == currException:
+      when true:
+        writeToStdErr "egotCurrentException() -> nil\n"
+        {.emit: """
```
Comment (Araq) [inline lib/system/excpt.nim]: This hack at least (!) needs a very good comment.

File: lib/system/excpt.nim [hunk 4]
```diff
@@ -474,12 +474,34 @@ when defined(endb):
 when defined(cpp) and appType != "lib" and
     not defined(js) and not defined(nimscript) and
     hostOS != "standalone" and not defined(noCppExceptions):
+  proc verboseTerminate() {.importc: "__gnu_cxx::__verbose_terminate_handler", header: "<exception>".}
```
Comment (alaviss) [inline lib/system/excpt.nim]: I don't think the stdlib should depends on libstdc++, given that there are macosx and windows targets.

File: lib/system/excpt.nim (hunk 3 above)
Comment (cooldome) [reply lib/system/excpt.nim]: There is no need for emit here, here is example how to do it: https://github.com/nim-lang/Nim/blob/devel/tests/exception/tcpp_imported_exc.nim

```nim
type 
  std_exception* {.importcpp: "std::exception", header: "<exception>".} = object

proc what(ex: std_runtime_error): cstring {.importcpp: "((char *)#.what())".}

try:
  myexception()
except std_exception as ex:
  echo $ex.what()
```

File: lib/system/excpt.nim (hunk 4 above)
Comment (pb-cdunn) [reply lib/system/excpt.nim]: `verboseTerminate()` is here only to show what is possible. `std::get_terminate()` would be better, and it's standard C++11. We could test a macro for C++11. Delegating to the normal terminate_handler is definitely what the C++ coder would expect.

The question is what to do for C++ < 11. I'm fine with `quit 1` in that case. I'm also fine with re-throwing, catching, and printing `what()` instead.

File: lib/system/excpt.nim (hunk 3 above)
Comment (pb-cdunn) [reply lib/system/excpt.nim]: Wouldn't we still need to "emit" a re-throw for `myexception()`?

**Comment by pb-cdunn** [comment]
```
Error: execution of an external compiler program 'g++ -c  -w -w -fpermissive  -I/home/travis/build/nim-lang/Nim/lib -I/home/travis/build/nim-lang/Nim/tests/cpp -o /home/travis/build/nim-lang/Nim/nimcache/tests/cpp/t8241.nim_f6f87c9fdcf8b3c3f07f93f1ee8712c9/stdlib_system.cpp.o /home/travis/build/nim-lang/Nim/nimcache/tests/cpp/t8241.nim_f6f87c9fdcf8b3c3f07f93f1ee8712c9/stdlib_system.cpp' failed with exit code: 1

/home/travis/build/nim-lang/Nim/nimcache/tests/cpp/t8241.nim_f6f87c9fdcf8b3c3f07f93f1ee8712c9/stdlib_system.cpp: In function â€˜void systemInit000()â€™:
/home/travis/build/nim-lang/Nim/nimcache/tests/cpp/t8241.nim_f6f87c9fdcf8b3c3f07f93f1ee8712c9/stdlib_system.cpp:6962:42: error: â€˜get_terminateâ€™ is not a member of â€˜stdâ€™
  oldTerminate_39bIE7CEicaKH9a7ZbsDlMkA = std::get_terminate();
                                          ^
```
The above failed because it was compiled for C++ < 11. We can add a macro to check that, but we need to decide what to do in that case. It might be simplest to use the "emit" code to re-throw, trap, and print `what()`.

**Comment by pb-cdunn** [comment]
To be clear, this is a fix for a bug in Nim standard library code. An exception thrown in C++ should not lead to a seg-fault in standard Nim-generated code.

**Comment by pb-cdunn** [comment]
Ping. (Simple bug; simple fix.)

**Comment by pb-cdunn** [comment]
Note: Because of #10898, we need a new release of Nim pretty soon for C++ code generation. (I'm forced to work off `devel` within my company.) I'd also really like a fix for this separate problem in the next release.

To summarize: When an exception is thrown in C++ and not trapped, Nim library code crashes. That is a bug in Nim. Not horrendous, but also not hard to fix.

**Comment by cooldome** [comment]
I will take over and prepare alternative fix this evening

**Comment by cooldome** [comment]
alternative fix went it, closing

**Pull request closed without merging**